### ISSUE TITLE
Can the docs point out the pros of browserify?

### ISSUE THREAD

**Issue by taoeffect** [vuejs/vuejs.org#252]
On the [documentation's recommendation](http://vuejs.org/guide/application.html#Single_File_Components) I started out developing our site+app using Webpack, and spent a [long time](https://github.com/okTurtles/group-income/wiki/Architecture-Notes) researching Webpack.

I came to the conclusion that for our setup Webpack really didn't work well for these reasons:

- webpack discourages use of grunt, pushes itself as "all-in-one system" but doesn't deliver as nicely as grunt+browserify
- backwards .html support: https://caurea.org/2015/09/16/webpack-is-not-a-static-site-generator.html
- browserify allows simple use of `require` and seems to have better node compatibility: https://mattdesl.svbtle.com/browserify-vs-webpack

The [poor HTML support](https://caurea.org/2015/09/16/webpack-is-not-a-static-site-generator.html) was a deal breaker for us, and in the end I am really happy with the flexibility and power that grunt offers. Our grunt file makes it possible (and likely simple compared to Webpack) to run a single command: `grunt dev`, and have multiple servers launch (backend API server, browser livereload listener for when files are updated, and the frontend server for displaying the site).

So I really feel that if the docs are going to be pushing the pros of Webpack, they should also at least mention the advantages of browserify, just for fairness and completeness, as it would be helpful to others.

**Comment by yyx990803** [comment]
What prevents you from using Grunt with Webpack? You can write a simple grunt task or node script to launch servers, why does it have anything to do with Webpack vs. Browserify?

The HTML problem wasn't explained clearly either. How does Browserify handles this better? If you need Grunt+Browserify to do it, then why can't you do it with Grunt+Webpack?

Inline require overloading is optional, in fact it's very rare to use it in application code. Not supporting it doesn't really make it a pro on Browserify's side.

**Comment by taoeffect** [comment]
> What prevents you from using Grunt with Webpack? You can write a simple grunt task or node script to launch servers, why does it have anything to do with Webpack vs. Browserify?

Webpack pushes itself as an all-in-one solution, with its own crazy stuff [like this](https://webpack.github.io/docs/webpack-dev-server.html), whereas Browserify does not. Browserify does not attempt to re-invent the wheel as webpack does for a lot of things.

Maybe you can use Webpack with grunt, but I doubt for most people it will make as much sense since they will have to fight between choosing the webpack version of doing something and the normal grunt way of doing it. Browserify doesn't have this problem.

Things that *grunt should be doing* (like running servers), the Webpack docs and tutorials [tell you to run things like `webpack-dev-server` from your `package.json`](https://github.com/AriaFallah/WebpackTutorial/tree/master/part1#the-development-server).

Webpack pushes its own crazy `webpack.config.dev.js` file as a replacement for `Gruntfile.js`, encouraging people to do [stuff like this](https://github.com/AriaFallah/WebpackTutorial/tree/master/part1#the-development-server):

```js
// webpack.config.dev.js
var path = require('path')
var webpack = require('webpack')
var HtmlWebpackPlugin = require('html-webpack-plugin')

module.exports = {
  devtool: 'cheap-eval-source-map',
  entry: [
    'webpack-dev-server/client?http://localhost:8080',
    'webpack/hot/dev-server',
    './src/index'
  ],
  output: {
    path: path.join(__dirname, 'dist'),
    filename: 'bundle.js'
  },
  plugins: [
    new webpack.HotModuleReplacementPlugin(),
    new HtmlWebpackPlugin({
      template: './src/index.html'
    })
  ],
  module: {
    loaders: [{
      test: /\.css$/,
      loaders: ['style', 'css']
    }]
  },
  devServer: {
    contentBase: './dist',
    hot: true
  }
}
```

Users are repeatedly encouraged to do stuff like that—stuff that grunt should be doing if they are using grunt. Webpack has lots of extraneous stuff that you have to spend time learning, and it's unclear why things are there, like what is the reason there's a `HtmlWebpackPlugin`? Browserify has no such thing, it's totally unnecessary. `grunt-contrib-connect` works great, there's no need for `webpack-dev-server`, `webpack-dev-middleware`, etc.

Using webpack with grunt, I just wonder, "Why bother? I'll be either dealing with a confusing mess, or will be using 5% of webpack and ignoring 95% of the tutorials and documentation out there."

> The HTML problem wasn't explained clearly either. How does Browserify handles this better?

Browserify just created javascript. 99% of webpack tutorials I came across, as well as its documentation, does [things like this](https://github.com/AriaFallah/WebpackTutorial/tree/master/part1#bundling):

```js
// index.js
require('./styles.css')
require('./UIStuff.js')
require('./APIStuff.js')

// UIStuff.js
var React = require('React')
React.createClass({
  // stuff
})

// APIStuff.js
var fetch = require('fetch') // fetch polyfill
fetch('https://google.com')
```

The entire philosophy of webpack seems to be to put as much as possible into JavaScript files as opposed to .html files.

Instead of having CSS be included form the HTML file, it has it pushed through webpack. Instead of a site with multiple HTML files, it expects you to have just one.

Browserify does not push this sort of pattern. Learning browserify makes sense: "You have this thing to generate JS out of JS, and you include that JS from your HTML files." Learning Webpack is like: "You generate JS out of your CSS, JPGs, and everything else except for HTML files, and we don't really think about HTML files at all."

I literally could not find *a single* tutorial explaining how to use Webpack with normal web development that designers are used to doing. You know—regular HTML files. You are more than welcome to point me to one.

> Inline require overloading is optional, in fact it's very rare to use it in application code. Not supporting it doesn't really make it a pro on Browserify's side.

Browserify supports much of Node's API and integrates nicely with it. I don't think Webpack does. That's a clear pro for browserify.

So in summary, the pro's for browserify that should be listed are:

- Clear separation of concerns, does not re-invent 10 wheels and then say, "well, you can use me with 5 of those wheels I re-invented anyway if you want"
- Easier for pick up
- Better Node support
- More friendly for designers who are used to generating normal HTML

**Comment by taoeffect** [comment]
As in one of [the links](https://caurea.org/2015/09/16/webpack-is-not-a-static-site-generator.html) in the OP:

> It's much simpler to set up browserify with a few plugins (such as babel or typescript) and produce a single output file. The file you then reference from your HTML file.

- Browserify is simpler. It focuses on one thing and it does it well. It's super clear how to use it with multiple HTML files (unlike Webpack).
- Grunt is simpler. It focuses on one thing and it does it better than Webpack.

**Comment by taoeffect** [comment]
Just want to be clear: I am not trying to say that Webpack is bad or anything like that. It seems like it does a lot of things better than browserify, and some things that browserify doesn't do. As far as a JS bundler goes it seems to be excellent. However, that doesn't make it "the best choice" for everyone. It would be great if the documentation reflected the reality that for some people, browserify may be a better choice for various reasons.

**Comment by yyx990803** [comment]
I think you are totally missing the point about Webpack: it serves a totally different purpose than grunt. Webpack is a **module bundler** and Grunt is a **task runner**. Why can't you just use grunt to spawn a process that runs `webpack-dev-server`? Comparing Webpack vs. Grunt is like comparing apples and oranges.

In fact, when using webpack, most of the time the replacement for grunt is **npm scripts**, in which you can just call arbitrary node scripts to do what grunt does.

You also don't seem to understand why Webpack pushes to put everything in JavaScript, because that is what allows us to have **a single dependency graph** between all the assets. Your html depends on your JS, your JS depends on other JS, your JS components depends on CSS and images... by requiring them from JavaScript, you are naturally creating this dependency graph using relative paths during development, without having to worry about rewriting them (e.g. with hash fingerprints) when deploying for production.

Webpack's config can be daunting, but that's why it is intended for "large scale applications": you pay the initial cost of configuration for long term productivity instead of taking the path of least resistance and missing out later.

Note that in the docs, the whole section is dedicated to "large scale applications", and personally I'd never recommend building large scale applications using multiple static html pages with sprinkled assets. For the specific use case discussed in that section, I think Webpack is objectively better.

**Comment by taoeffect** [comment]
In order to use Vuejs .vue files you need to use browserify or webpack, correct?

If that's correct, then there is a use case for a "non-large-scale-app" with .vue files, correct?

If you agree with those two, then the only conclusion that you can make is that there is a legitimate use case for using browserify over webpack, where webpack just doesn't make sense.

What does leaving the documentation as-is accomplish except mislead a good portion of your readers?

I have not heard yet a good answer on this from you. You call your documentation objective, but it's clearly not. There is no question about that.

- Fact: There are use cases for VueJS where browserify is the preferred option. Our project is one of them.
- Fact: grunt cannot be cleanly by npm scripts. If you disagree, then prove it. Show us how to convert [this Gruntfile](https://gist.github.com/taoeffect/8a8a98d868bc0596c0dc) to npm scripts.

So, you can certainly have your document be opinionated and say things like, "In our experience, Webpack is better for large scale applications", however, it should also say that "For other VueJS use cases, browserify might be a better fit", or something like that (and maybe link to this thread or one of the links I posted).

Is there any issue with that? I can't think of one.

Would such a change be OK? If not, can you please explain your reasoning?

**Comment by yyx990803** [comment]
Yeah, I'd be okay with that if your issue opened with that suggestion directly - however your original post didn't clearly explain that and instead was mainly piling on what you didn't like about Webpack, which leads to me thinking that you simply don't understand how Webpack works.

I apologize if I sounded blunt, and I'd be happy to take a PR on what you proposed in your last comment.

**Comment by taoeffect** [comment]
Awesome, thanks so much @yyx990803! Your reply has restored my faith in OSS project maintainers, and I apologize for any lack of clarity on my part! Vue.js is awesome btw, you've really done an amazing job on it, the labor of love shows.

🙇

**Comment by Silur** [comment]
A good example or the pros of browserify is vue+electron.
There are several NPM modules which will die miserably within webpack, for eg websockets, no god can help you to get them work within webpack.
However I've never seen an npm package to die under browserify :)
